File: app/templates/app-java/src/main/java/util/gson/CurrencyUnitTypeConverter.java
package <%= appPackage %>.util.gson;

import java.lang.reflect.Type;

import org.joda.money.CurrencyUnit;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

public class CurrencyUnitTypeConverter implements JsonSerializer<CurrencyUnit>, JsonDeserializer<CurrencyUnit> {
	@Override
	public JsonElement serialize(CurrencyUnit src, Type srcType, JsonSerializationContext context) {
		return new JsonPrimitive(serialize(src));
	}

	@Override
	public CurrencyUnit deserialize(JsonElement json, Type type, JsonDeserializationContext context) throws JsonParseException {
		return deserialize(json.getAsString());
	}

	CurrencyUnit deserialize(String jsonString) {
		return CurrencyUnit.of(jsonString);
	}

	String serialize(CurrencyUnit src) {
		return src.getCurrencyCode();
	}
}

File: app/templates/app-kotlin/src/main/java/util/gson/CurrencyUnitTypeConverter.kt
package <%= appPackage %>.util.gson

import java.lang.reflect.Type

import org.joda.money.CurrencyUnit

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer

class CurrencyUnitTypeConverter : JsonSerializer<CurrencyUnit>, JsonDeserializer<CurrencyUnit> {

    override fun serialize(src: CurrencyUnit, srcType: Type, context: JsonSerializationContext): JsonElement {
        return JsonPrimitive(serialize(src))
    }

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, type: Type, context: JsonDeserializationContext): CurrencyUnit {
        return deserialize(json.asString)
    }

    internal fun deserialize(jsonString: String): CurrencyUnit {
        return CurrencyUnit.of(jsonString)
    }

    internal fun serialize(src: CurrencyUnit): String {
        return src.currencyCode
    }
}
